File: apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/ResolveDelegate.java
package com.apollographql.apollo.internal.response;

import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.internal.Optional;

import java.util.List;

public interface ResolveDelegate<R> {

  void willResolveRootQuery(Operation operation);

  void willResolve(ResponseField field, Operation.Variables variables);

  void didResolve(ResponseField field, Operation.Variables variables);

  void didResolveScalar(Object value);

  void willResolveObject(ResponseField objectField, Optional<R> objectSource);

  void didResolveObject(ResponseField objectField, Optional<R> objectSource);

  void didResolveList(List array);

  void willResolveElement(int atIndex);

  void didResolveElement(int atIndex);

  void didResolveNull();
}

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResolveDelegate.kt
package com.apollographql.apollo3.api.internal

import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.ResponseField

interface ResolveDelegate<R> {
  fun willResolveRootQuery(operation: Operation<*>)
  fun willResolve(field: ResponseField, variables: Operation.Variables, value: Any?)
  fun didResolve(field: ResponseField, variables: Operation.Variables)
  fun didResolveScalar(value: Any?)
  fun willResolveObject(objectField: ResponseField, objectSource: R?)
  fun didResolveObject(objectField: ResponseField, objectSource: R?)
  fun didResolveList(array: List<*>)
  fun willResolveElement(atIndex: Int)
  fun didResolveElement(atIndex: Int)
  fun didResolveNull()
}
